src/token/cose/key.rs

```diff
@@ -0,0 +1,444 @@
+        match &crv {
+            EllipticCurve::Assigned(
+                iana::EllipticCurve::P_256
+                | iana::EllipticCurve::P_384
+                | iana::EllipticCurve::P_521
+                | iana::EllipticCurve::Secp256k1,
+                // TODO these are also part of the IANA registry, and should be added to coset
+                // | iana::EllipticCurve::BrainpoolP256r1
+                // | iana::EllipticCurve::BrainpoolP320r1
+                // | iana::EllipticCurve::BrainpoolP384r1
+                // | iana::EllipticCurve::BrainpoolP512r1
+            ) => {}
+            // Any other assigned values must not be encoded as EC2 values.
+            v @ EllipticCurve::Assigned(_) => {
+                return Err(CoseCipherError::KeyTypeCurveMismatch(
+                    KeyType::Assigned(iana::KeyType::EC2),
+                    v.clone(),
+                ))
+            }
+            // Anything else is private use/custom, and we assume that the caller knows what they
+            // are doing.
+            _v => {}
+        }
+
+        // Parse parameters d and x (private key and part of the public key), must be of type
+        // bstr.
+        let d = find_param_by_label(&Label::Int(iana::Ec2KeyParameter::D.to_i64()), &key.params)
+            .map(|v| match v.as_bytes() {
+                None => Err(CoseCipherError::InvalidKeyParam(
+                    iana::Ec2KeyParameter::D.into(),
+                    v.clone(),
+                )),
+                Some(b) => Ok(b.as_slice()),
+            })
+            .transpose()?;
+        let x = find_param_by_label(&Label::Int(iana::Ec2KeyParameter::X.to_i64()), &key.params)
+            .map(|v| match v.as_bytes() {
+                None => Err(CoseCipherError::InvalidKeyParam(
+                    iana::Ec2KeyParameter::X.into(),
+                    v.clone(),
+                )),
+                Some(b) => Ok(b.as_slice()),
+            })
+            .transpose()?;
+
+        // Parse parameters y (other half of public key), is either a bstr containing the Y
+        // coordinate or a boolean indicating the sign.
+        let y = find_param_by_label(&Label::Int(iana::Ec2KeyParameter::Y.to_i64()), &key.params);
+        let (y, sign) = match y {
+            None => (None, None),
+            Some(Value::Bytes(b)) => (Some(b.as_slice()), None),
+            Some(Value::Bool(b)) => (None, Some(b.clone())),
+            Some(value) => {
+                return Err(CoseCipherError::InvalidKeyParam(
+                    iana::Ec2KeyParameter::Y.into(),
+                    value.clone(),
+                ))
+            }
+        };
+
+        // For public keys, X and Y must be set, for private keys, at least D must be set.
+        if d.is_none() && !(x.is_some() && (y.is_some() || sign.is_some())) {
+            return Err(CoseCipherError::MissingKeyParam(
+                iana::Ec2KeyParameter::D.into(),
```

Comment:
Hm, no, I'm just not a huge fan of giving "wrong" error messages, as that can be slightly confusing when knee-deep in some cryptography debugging session. I see the point about the many error variants that'd bloat the code, though. Maybe it's better to remove the parameter here altogether?

Reply:
I fear that that might make the issue more confusing to debug, as you'd then only get a "MissingKeyParam" error without a clue as to what parameter might be missing.

Maybe we could provide a `Vec` of missing parameters instead (in that case, including all three parameters as missing)? 
That way, the result would at least be _less_  incorrect, although it might falsely indicate to the user that all three parameters are required for this operation to succeed.

Reply:
> I fear that that might make the issue more confusing to debug, as you'd then only get a "MissingKeyParam" error without a clue as to what parameter might be missing.

Yes, but I think I'd rather get an error that tells me versus an error that sets me on the wrong trail. What you propose in the second paragraph is a reasonable compromise, though—as long as the documentation indicates that at least one of the `Vec` members is missing.